tuningFork.c

```diff
@@ -203,93 +186,61 @@ void genFile(double freq, double duration, char* fname)
                 /* 40 */ 0x74, 0x61,   B4,   B3,   B2,   B1};
 
+    /////////////////////////
+    // Generate data subchunk
+    /////////////////////////
 
-    //////////////////////
-    // Calculate samples
-    //////////////////////
-
-
-    const int NUM_PERIODS = 400;
-    int16_t sample; // designates value of wave func at that point
     int max_volume = 0x7FFF;
-    double omega = 2*PI*freq; // cuts down on calculation time
-
-    int i;
-
-    double period = 1/freq; // in seconds
-    int tableSize = NUM_PERIODS * SAMPLE_RATE * period; // in samples
-    uint16_t* sampleTable = malloc(sizeof(uint16_t) * tableSize);
-
-    int dataLoopLimit = SAMPLE_RATE * duration;
-    uint16_t* dataArr = malloc(sizeof(uint16_t) * 2 * dataLoopLimit);
+    int numSamples = SAMPLE_RATE * duration;
 
     // Generate a lookup table for samples
-    //int nThreads = omp_get_num_procs();
-    //omp_set_num_threads(nThreads);
-
-    double omegaT = 0; // omega * time
-    double deltaOmegaT = (double)omega/SAMPLE_RATE;
-
-
-    for (i = 0; i < tableSize; i++)
-    {
-        //fprintf(stderr, "omegaT is %lf\n", omegaT);
-        sample = max_volume * sin(omegaT);
-        //sample = max_volume * sin(omega * (double)i/SAMPLE_RATE);
-
-        sampleTable[i] = sample;
-
-        omegaT += deltaOmegaT; // gives next value to evaluate sin at
+    const int lookupTableSize = 8192;  // Size in samples
+    const int quarterTableSize = lookupTableSize >> 2;
+    int16_t* lookupTable = malloc(sizeof(int16_t) * lookupTableSize);
+
+    // Generate lookup table for 1/4 period of a sine wave
+    // We can evaluate any value over an entire period using just the first quadrant
+    for (int i = 0; i < quarterTableSize; i++) {
+        lookupTable[i] = max_volume * sin((PI / 2) * ((double)i / quarterTableSize));
+        lookupTable[(quarterTableSize * 2) - i - 1] = lookupTable[i];
+        lookupTable[i + quarterTableSize * 2] = ~lookupTable[i];
+        lookupTable[(quarterTableSize * 4) - i - 1] = ~lookupTable[i];
     }
 
-
-    int periodIndex = 0; // the index within the current period relative to
-                         // the start of the period
-
-    // Populate sample array from lookup table
-    for (i = 0; i < dataLoopLimit; i++)
-    {
-        //periodIndex = i % tableSize;
-
-        sample = sampleTable[periodIndex];
-
-        /* write two samples to array */
-        int j = i*2;
-        dataArr[j] = sample;
-        dataArr[j+1] = sample;
-
-        periodIndex++;
-        if ( periodIndex >= tableSize)
-        {
-            // reset the index
-            periodIndex = 0;
-        }
+    const double phaseIncrement = (double)(freq / (2.0 * SAMPLE_RATE)) * lookupTableSize;
+    double phase = 0.0;
+    int phase_i = 0;
+    int16_t* dataSubChunk = malloc(sizeof(int16_t) * numSamples * 2);
+
+    for (int i = 0; i < numSamples; ++i) {
+        phase_i = (int)phase;
+        int j = i * 2;
```

Comment:
Could we move j out of the loop and increment it by 2 at each iteration instead? That would run more efficiently I think